test/helpers/erc4337.js

```diff
@@ -83,27 +83,41 @@ class SmartAccount extends ethers.BaseContract {
 }
 
 class ERC7702SmartAccount extends SmartAccount {
-  constructor(instance, delegate, env) {
+  constructor(instance, authorization, env) {
     super(instance, undefined, env);
-    this.delegate = delegate;
+    this.authorization = authorization;
   }
 
   async deploy() {
-    await ethers.provider.getCode(this.delegate).then(code => setCode(this.target, code));
+    // hardhat signers from @nomicfoundation/hardhat-ethers do not support type 4 txs.
+    // so we rebuild it using "native" ethers
+    await ethers.Wallet.fromPhrase(config.networks.hardhat.accounts.mnemonic, ethers.provider).sendTransaction({
```

Comment:
Isn't this tied to the first account? I tried to make `AccountERC7702WithModules` to revert unsuccessfully, but I think we should do something like:

```js
await ethers.Wallet.fromPhrase(this.findAccount(), ethers.provider).sendTransaction(...)

findAccount() {
  for (const account of config.networks.hardhat.accounts) {
    console.log(account);
    if (account.address === this.address) {
      return account;
    }
  }
}
```

Reply:
config.network.hardhat.accounts is not an array. It's an object that contains a mnemonic from which all signers are derived. It does not contain an address.

And yes, this code uses the first one of the bunch.

Reply:
Also, if we make the EOA relay its own delegation, we get into nonce issues that I'd rather avoid.

Reply:
Basically:
- any account can relay
- account needs gas
- account objects needs to support type4 txs (hardhat-ether) signer do not
- if the account used is the signer of the delegate, we need to fix the nounces.

So my choice was to take the first "canonical" account from the hardhat config and instantiate it as a signer that supports type 4 txs

Reply:
Right! So the "first canonical account" just relays. I forgot we reserved the first account, so that explains why I couldn't make the test fail.

Makes sense then

Reply:
First account is not reserved.

I'm not sure what you think could cause any of that to fail.

Reply:
> If I remember correctly, the first account is no longer reserved.

Also right, I thought it was still around and now I just realized that an EOA can send a transaction with another's EOA authorization 🤯 didn't realize that but it's extremely convenient

Reply:
That is why you need the authorizations to be signed. Authorizations are like a meta transaction.

In particular, and ERC-4337 relayer can include the authorization with the user operation so that the 4337+7702 logic is deployed and used in a single operation